graylog2-server/src/main/java/org/graylog2/database/PaginatedDbService.java

```diff
@@ -153,25 +152,18 @@ public int delete(String id) {
                                                                      DBSort.SortBuilder sort,
                                                                      int page,
                                                                      int perPage) {
-        final AtomicInteger total = new AtomicInteger(0);
+        final long total = streamQueryWithSort(query, sort).filter(filter).count();
```

Comment:
Please make sure to close the stream because it's backed by a MongoDB cursor. (as documented in the `streamQueryWithSort()` javadoc)

Reply:
Thanks for the reminder, I was actually wondering about the relevance of that comment, as it should be taken care of the `onClose` handler which is [added to the stream upon instantiation](https://github.com/Graylog2/graylog2-server/blob/master/graylog2-server/src/main/java/org/graylog2/database/PaginatedDbService.java#L213), or am I wrong?

Reply:
The `onClose` handler is called when a stream is closed by calling `.close()` or using something like try-with-resources. At least to my knowledge. :smiley: 

The Java Stream docs say:

> Streams have a BaseStream.close() method and implement AutoCloseable, but nearly all stream instances do not actually need to be closed after use. **Generally, only streams whose source is an IO channel (such as those returned by Files.lines(Path, Charset)) will require closing.**

Source: https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html

I think the last part in that excerpt is relevant to our use case since the returned stream is backed by a MongoDB cursor object.

Reply:
That said, the MongoDB server will close the cursor automatically at some point:

> Instructs the server to close a cursor and free associated server resources. The server will automatically close cursors that have no remaining results, as well as cursors that have been idle for a period of time and lack the cursor.noCursorTimeout() option.

Source: https://docs.mongodb.com/manual/reference/method/cursor.close/

So while it's not a huge issue not closing the cursor, I think we should always try to close the cursor if we can to avoid having dangling cursors.

Reply:
✅